File: packages/provide-uterm-csharp/src/Provide.Uterm/Hub/Store.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

namespace Provide.Uterm.Hub;

public enum OwnershipPublicationExpectation
{
    RestHeld,
    DashboardHeld,
    Released,
}

public sealed record OwnershipPublicationToken
{
    private OwnershipPublicationToken(
        string workerId,
        long ownershipVersion,
        OwnershipPublicationExpectation expectation,
        bool enabled,
        string? publishedOwner,
        string? restHijackId,
        string? restOwner,
        object? dashboardOwner)
    {
        WorkerId = workerId;
        OwnershipVersion = ownershipVersion;
        Expectation = expectation;
        Enabled = enabled;
        PublishedOwner = publishedOwner;
        RestHijackId = restHijackId;
        RestOwner = restOwner;
        DashboardOwner = dashboardOwner;
    }

    public string WorkerId { get; }
    public long OwnershipVersion { get; }
    public OwnershipPublicationExpectation Expectation { get; }
    public bool Enabled { get; }
    public string? PublishedOwner { get; }
    public string? RestHijackId { get; }
    public string? RestOwner { get; }
    public object? DashboardOwner { get; }

    public static OwnershipPublicationToken RestHeld(
        string workerId,
        long ownershipVersion,
        string hijackId,
        string owner) => new(
            workerId,
            ownershipVersion,
            OwnershipPublicationExpectation.RestHeld,
            true,
            owner,
            hijackId,
            owner,
            null);

    public static OwnershipPublicationToken DashboardHeld(
        string workerId,
        long ownershipVersion,
        object owner,
        string? publishedOwner = null) => new(
            workerId,
            ownershipVersion,
            OwnershipPublicationExpectation.DashboardHeld,
            true,
            publishedOwner,
            null,
            null,
            owner);

    public static OwnershipPublicationToken Released(
        string workerId,
        long ownershipVersion) => new(
            workerId,
            ownershipVersion,
            OwnershipPublicationExpectation.Released,
            false,
            null,
            null,
            null,
            null);
}

/// <summary>Input-buffer + lifecycle/policy helper service.</summary>
public sealed class StateStore
{
    private readonly WorkerRegistry _registry;
    private readonly object _lock;
    private readonly IClock _clock;
    private readonly int _maxBufferChars;
    private readonly Action<string, int>? _onMetric;
    private readonly Action<string, bool, string?>? _onHijackChanged;
    private readonly object _buffersGate = new();
    private readonly Dictionary<object, string> _inputBuffers = new();
    private readonly object _hijackNotificationGatesLock = new();
    private readonly Dictionary<string, object> _hijackNotificationGates =
        new(StringComparer.Ordinal);

    public StateStore(
        WorkerRegistry registry,
        object sharedLock,
        IClock? clock = null,
        int maxBufferChars = 40000,
        Action<string, int>? onMetric = null,
        Action<string, bool, string?>? onHijackChanged = null)
    {
        _registry = registry;
        _lock = sharedLock;
        _clock = ClockUtil.OrDefault(clock);
        _maxBufferChars = maxBufferChars;
        _onMetric = onMetric;
        _onHijackChanged = onHijackChanged;
    }

    public static int ClampLease(int leaseS)
    {
        if (leaseS < 1) return 1;
        if (leaseS > 14400) return 14400;
        return leaseS;
    }

    public bool HasValidRestLease(WorkerTermState st) =>
        st.HijackSession is not null && st.HijackSession.LeaseExpiresAt > _clock.Monotonic();

    public bool IsDashboardHijackActive(WorkerTermState st)
    {
        if (st.HijackOwner is null) return false;
        if (st.HijackOwnerExpiresAt is null) return true;
        return st.HijackOwnerExpiresAt.Value > _clock.Monotonic();
    }

    public bool IsHijacked(WorkerTermState st) =>
        IsDashboardHijackActive(st) || HasValidRestLease(st);

    public void Metric(string name, int value = 1) => _onMetric?.Invoke(name, value);

    public void NotifyHijackChanged(string workerId, bool enabled, string? owner)
    {
        OwnershipPublicationToken? token;
        lock (_lock)
        {
            var st = _registry.Get(workerId);
            token = st is null
                ? null
                : CurrentPublicationToken(workerId, st, enabled, owner);
        }
        if (token is not null) NotifyHijackChanged(token);
    }

    /// <summary>
    /// Publish a captured ownership transition only while its generation,
    /// ownership kind, and owner identity still match current state. The
    /// per-worker notification gate establishes one callback order for all
    /// acquisition, release, force-release, expiry, and disconnect publishers.
    /// </summary>
    public bool NotifyHijackChanged(OwnershipPublicationToken token)
    {
        lock (HijackNotificationGate(token.WorkerId))
        {
            lock (_lock)
            {
                var st = _registry.Get(token.WorkerId);
                if (st is null
                    || st.HijackOwnershipVersion != token.OwnershipVersion
                    || !MatchesExpectedOwnership(st, token))
                {
                    return false;
                }
            }

            _onHijackChanged?.Invoke(
                token.WorkerId, token.Enabled, token.PublishedOwner);
            return true;
        }
    }

    private OwnershipPublicationToken? CurrentPublicationToken(
        string workerId,
        WorkerTermState st,
        bool enabled,
        string? owner)
    {
        if (!enabled)
        {
            return OwnershipPublicationToken.Released(
                workerId, st.HijackOwnershipVersion);
        }
        if (HasValidRestLease(st) && st.HijackSession is { } rest)
        {
            return OwnershipPublicationToken.RestHeld(
                workerId, st.HijackOwnershipVersion, rest.HijackId, rest.Owner);
        }
        if (IsDashboardHijackActive(st) && st.HijackOwner is { } dashboard)
        {
            return OwnershipPublicationToken.DashboardHeld(
                workerId, st.HijackOwnershipVersion, dashboard, owner);
        }
        return null;
    }

    private bool MatchesExpectedOwnership(
        WorkerTermState st,
        OwnershipPublicationToken token) => token.Expectation switch
    {
        OwnershipPublicationExpectation.RestHeld =>
            HasValidRestLease(st)
            && st.HijackSession is { } rest
            && rest.HijackId == token.RestHijackId
            && rest.Owner == token.RestOwner,
        OwnershipPublicationExpectation.DashboardHeld =>
            IsDashboardHijackActive(st)
            && ReferenceEquals(st.HijackOwner, token.DashboardOwner),
        OwnershipPublicationExpectation.Released =>
            !IsDashboardHijackActive(st) && !HasValidRestLease(st),
        _ => false,
    };

    private object HijackNotificationGate(string workerId)
    {
        lock (_hijackNotificationGatesLock)
        {
            if (!_hijackNotificationGates.TryGetValue(workerId, out var gate))
            {
                gate = new object();
                _hijackNotificationGates[workerId] = gate;
            }
            return gate;
        }
    }

    public void TouchActivity(string workerId)
    {
        lock (_lock)
        {
            var st = _registry.Get(workerId);
            if (st is not null)
            {
                st.LastActivityAt = _clock.Monotonic();
            }
        }
    }

    public WorkerTermState GetOrCreate(string workerId)
    {
        lock (_lock)
        {
            var st = _registry.Get(workerId);
            if (st is null)
            {
                st = new WorkerTermState();
                _registry.Put(workerId, st);
            }

            return st;
        }
    }

    public (string Command, bool Ok) BufferAndGetCommand(object ws, string data)
    {
        lock (_buffersGate)
        {
            var buf = (_inputBuffers.TryGetValue(ws, out var existing) ? existing : "") + data;
            if (buf.Length > _maxBufferChars)
            {
                _inputBuffers.Remove(ws);
                return ("", false);
            }

            if (buf.Contains('\r') || buf.Contains('\n'))
            {
                _inputBuffers.Remove(ws);
                return (buf, true);
            }

            _inputBuffers[ws] = buf;
            return ("", false);
        }
    }

    public int Shutdown() => 0;
}

File: packages/provide-uterm-ts/src/hub/store.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Input buffering, worker lifecycle and the hijack-state predicates.
 *
 * Port of the Python module `provide.uterm.server.bridge.hub.store` and the Go
 * package `hub` (`store.go`).
 *
 * The reference reaches these through a back reference to the composing hub;
 * this takes the same surface as explicit options, which is what the Go port
 * does and what makes the service testable without standing up a hub.
 */

import { type Logger, noopLogger } from "../telemetry/index.ts";
import { type Connection, WorkerTermState } from "./models.ts";
import type { WorkerRegistry } from "./registry.ts";

/** Shortest lease the hub will grant, in seconds. */
export const LEASE_MIN_SECONDS = 1;

/**
 * Longest lease the hub will grant, in seconds — four hours.
 *
 * Matched to the WebSocket idle-reader timeout so a long operator hold is not
 * killed by whichever of the two fires first. The earlier one-hour cap was too
 * tight for multi-hour runs that drive one target hard while others idle.
 */
export const LEASE_MAX_SECONDS = 14400;

/**
 * Clamp a requested lease duration into the accepted range.
 *
 * Truncates toward zero first, matching the reference's `int()` coercion: a
 * request for 90.9 seconds grants 90, never 91.
 */
export function clampLease(leaseSeconds: number): number {
  return Math.max(LEASE_MIN_SECONDS, Math.min(Math.trunc(leaseSeconds), LEASE_MAX_SECONDS));
}

/** Construction options for {@link StateStore}. */
export interface StateStoreOptions {
  /** Worker table this store creates into and stamps activity on. */
  registry: WorkerRegistry<WorkerTermState>;
  /** Cap on a single browser's un-terminated input buffer. */
  maxBufferChars: number;
  /** Monotonic clock in seconds. */
  now?: () => number;
  /** Where callback failures are reported. */
  logger?: Logger;
  /** Metrics sink, if one is configured. */
  onMetric?: (name: string, value: number) => void;
  /** Notified when a worker's hijack state changes. */
  onHijackChanged?: (workerId: string, enabled: boolean, owner?: string) => void | Promise<void>;
}

/** Options for {@link StateStore.notifyHijackChanged}. */
export interface HijackChangedOptions {
  /** Whether the worker is now hijacked. */
  enabled: boolean;
  /** Who holds it, when it is now hijacked. */
  owner?: string;
}

/** Whether `value` is a promise, matching the reference's isawaitable check. */
function isAwaitable(value: unknown): value is Promise<unknown> {
  return typeof (value as { then?: unknown } | null | undefined)?.then === "function";
}

/** Monotonic seconds, from a clock that cannot jump backwards. */
function monotonicNow(): number {
  return performance.now() / 1000;
}

/** Input buffering, worker lifecycle and the hijack-state predicates. */
export class StateStore {
  readonly #registry: WorkerRegistry<WorkerTermState>;
  readonly #maxBufferChars: number;
  readonly #now: () => number;
  readonly #logger: Logger;
  readonly #onMetric: ((name: string, value: number) => void) | undefined;
  readonly #onHijackChanged: ((workerId: string, enabled: boolean, owner?: string) => void | Promise<void>) | undefined;
  /** Partial input lines, keyed by the browser that is typing them. */
  readonly #buffers = new Map<Connection, string>();

  constructor(options: StateStoreOptions) {
    this.#registry = options.registry;
    this.#maxBufferChars = options.maxBufferChars;
    this.#now = options.now ?? monotonicNow;
    this.#logger = options.logger ?? noopLogger;
    this.#onMetric = options.onMetric;
    this.#onHijackChanged = options.onHijackChanged;
  }

  /**
   * Accumulate input for `ws`, returning the line once one is terminated.
   *
   * A buffer that would exceed the cap is *discarded whole* rather than
   * truncated or flushed — an over-long paste vanishes instead of arriving in
   * pieces, which bounds what one connection can make the hub hold. The cap is
   * checked before the newline scan, so an over-long write is dropped even
   * when it contains a terminator.
   */
  bufferAndGetCommand(ws: Connection, data: string): string | undefined {
    const buffer = (this.#buffers.get(ws) ?? "") + data;
    if (buffer.length > this.#maxBufferChars) {
      this.#buffers.delete(ws);
      return undefined;
    }
    if (buffer.includes("\r") || buffer.includes("\n")) {
      this.#buffers.delete(ws);
      return buffer;
    }
    this.#buffers.set(ws, buffer);
    return undefined;
  }

  /** What `ws` currently has buffered, for inspection. */
  bufferedFor(ws: Connection): string | undefined {
    return this.#buffers.get(ws);
  }

  /**
   * Forget a browser's partial line.
   *
   * Called when the connection closes: an abandoned buffer would otherwise be
   * retained for the lifetime of the hub.
   */
  dropBuffer(ws: Connection): void {
    this.#buffers.delete(ws);
  }

  /** The state for `workerId`, creating and registering it if it is new. */
  getOrCreate(workerId: string): WorkerTermState {
    const existing = this.#registry.get(workerId);
    if (existing !== undefined) {
      return existing;
    }
    const created = new WorkerTermState({ now: this.#now });
    this.#registry.put(workerId, created);
    return created;
  }

  /**
   * Stamp `workerId` as active now.
   *
   * A worker that has gone is ignored rather than recreated: it can disconnect
   * between a frame arriving and being handled, and resurrecting its state
   * here would leak an entry the pruner never sees a connection for.
   */
  touchActivity(workerId: string): void {
    const state = this.#registry.get(workerId);
    if (state !== undefined) {
      state.lastActivityAt = this.#now();
    }
  }

  /**
   * Emit a named metric, if a sink is configured.
   *
   * The value is truncated toward zero, matching the reference's `int()`
   * coercion. A throwing sink is swallowed and logged: observability must not
   * be able to tear down the session it is reporting on.
   */
  metric(name: string, value = 1): void {
    if (this.#onMetric === undefined) {
      return;
    }
    try {
      this.#onMetric(name, Math.trunc(value));
    } catch (error) {
      this.#logger.warn({ metric: name, error }, "metric_callback_failed");
    }
  }

  /** Whether `state` holds an unexpired REST hijack lease. */
  hasValidRestLease(state: WorkerTermState): boolean {
    const session = state.hijackSession;
    return session !== undefined && session.leaseExpiresAt > this.#now();
  }

  /**
   * Whether `state` holds an active dashboard hijack.
   *
   * An owner carrying *no* expiry counts as active — a perpetual hold. Note
   * that {@link WorkerTermState.lease}'s `isDashboardActive` answers `false`
   * for that same state. Both readings are in the reference and they are
   * genuinely inconsistent; unifying them would change who is allowed to send
   * input, so each keeps its own answer and the corpus pins both.
   */
  isDashboardHijackActive(state: WorkerTermState): boolean {
    if (state.hijackOwner === undefined) {
      return false;
    }
    if (state.hijackOwnerExpiresAt === undefined) {
      return true;
    }
    return state.hijackOwnerExpiresAt > this.#now();
  }

  /** Whether `state` is under any active hijack, dashboard or REST. */
  isHijacked(state: WorkerTermState): boolean {
    return this.isDashboardHijackActive(state) || this.hasValidRestLease(state);
  }

  /**
   * Tell the configured subscriber that a worker's hijack state changed.
   *
   * An async subscriber is fired and forgotten — the caller is mid-transition
   * and must not block on it — but its rejection is caught and logged, since
   * an unhandled one would take the process down under Node's default policy.
   *
   * A subscriber that throws *synchronously* propagates, deliberately unlike
   * {@link metric}: the reference leaves this call unguarded.
   */
  notifyHijackChanged(workerId: string, options: HijackChangedOptions): void {
    const callback = this.#onHijackChanged;
    if (callback === undefined) {
      return;
    }
    const result = callback(workerId, options.enabled, options.owner);
    // Awaitability, not non-undefined: a synchronous subscriber may still
    // return something (the reference asks inspect.isawaitable for the same
    // reason).
    if (isAwaitable(result)) {
      result.catch((error: unknown) => {
        this.#logger.warn({ workerId, error }, "on_hijack_changed callback raised");
      });
    }
  }
}
